src/CustomMarshalers/src/System/Runtime/InteropServices/CustomMarshalers/ComDataHelpers.cs

```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace System.Runtime.InteropServices.CustomMarshalers
+{
+    internal static class ComDataHelpers
+    {
+        public static TView GetOrCreateManagedViewFromComData<T, TView>(object comObject, Func<T, TView> createCallback)
+        {
+            object key = typeof(TView);
+
+            if (Marshal.GetComObjectData(comObject, key) is TView managedView)
+            {
+                return managedView;
+            }
+            else
+            {
+                managedView = createCallback((T)comObject);
+                if (!Marshal.SetComObjectData(comObject, key, managedView))
```

Comment:
I assume there aren't any thread-safety issues here?

Reply:
Just verified in coreclr that there are no thread-safety issues here.